---
title: "StringDB Analysis"
params: 
    type: "any"
    bp: 5000
    minReadCount: 10
    normType: "poisson"
    sizeFactor: "mean"
    filter: 0.01
    byChrom: TRUE
    reuse: FALSE
---
Run on `r date()` with `r params$bp`bp size windows, `r params$normType` normalized with the `r params$sizeFactor` size factor. A minimum of `r params$minReadCount` CNVs were required for a segment to be recorded. Gene annotations were done as `r params$type` with the segment. 
```{r Load Relevant Libraries}
library(STRINGdb)
library(tidyverse)
library(Biostrings)
library(GenomicFeatures)
library(UpSetR)
library("chroma")
library(janitor)
library(knitr)
library(fuzzyjoin)
```

```{r Load Basic Data}
param_end <- paste(
  params$sizeFactor,
  params$filter,
  params$byChrom, params$minReadCount,"",
  sep = "-")
sample_data <- read.csv("~/2021-REU/CNV Analysis/partials/sample_data.tsv", sep = "") %>% mutate(sample_des = paste(str_remove_all(turtle, "_kidney"), tumor_sample_location))
annoloc <- paste("02-Annotation", params$type, params$bp, params$normType, param_end,sep = "/")
saveloc <- paste("04-StringDB", params$type, params$bp, params$normType,param_end, sep = "/")

partialsloc <- paste0("partials/stringdb/", paste(params$type, params$bp, params$normType,
  param_end,
  sep = "-"
))
string_db <- STRINGdb$new(version = "11", species = 9606, score_threshold = 150, input_directory = "04-StringDB/")
```


## Get Gene Annotations
This notebook assumes that files generated by previous Rmarkdown files are present at `r annoloc`, otherwise this will throw errors. 
```{r Get Descriptions}
getCNVs <- function(filename){
  cnv <- NULL
  out <- tryCatch(
    {
      load(filename)
    },
    error=function(cond) {
      message(paste("\nFile does not seem to be valid:", x))
      message("Here's the original error message:")
      message(cond)
      # Choose a return value in case of error
      return(NULL)
    },
    warning=function(cond) {
      message(paste("File caused a warning:", x))
      message("Here's the original warning message:")
      message(cond)
    }
  )
  return(cnv)
}

# get the filenames and actual anmes
file_list <- list.files(path=paste("partials/cnv", params$type, params$bp, params$normType,sep="/"), 
                        pattern=glob2rx(paste(params$sizeFactor, params$filter, 
                                              params$byChrom, params$minReadCount,"*_cnvs.gz",sep="-")),full.names = TRUE)
names <- str_split_fixed(str_split_fixed(str_remove_all(file_list, "gz|\\.|_cnvs"), "/",6)[,6],"-",5)[,5]
cnvs <- lapply(file_list, getCNVs)
names(cnvs) <- names
```

## StringDB Mapping 

```{r StringDB Mapping}
colors <- data.frame(TUMOUR=c(0.025, 0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4,8, 16, 32, 64),color= c('#5e9a4c', '#77b464', '#92cf7c', '#bfe4b2','#f1f1f1', '#f9cfe7', '#fea6d8', '#ef85be', '#dc68a1', '#c74983', '#aa316a', '#8d1952', '#6e0038'))

peformString <- function(x, cnvs) {
  string_db <- STRINGdb$new(version = "11", species = 9606, score_threshold = 150, input_directory = "04-StringDB/")
  if (is_null(cnvs[[x]])) { # guard against NULL
    return(NULL)
  }
  if (params$reuse & file.exists(paste0(saveloc, x, ".gz"))) {
    message(paste("File found for STRING on", x))
    load(paste0(saveloc, x, ".gz"))
    # genes <- multi_map_df(map, string_db$get_aliases(), "STRING_id", "STRING_id", "alias")
  } else {
    cnv_data <- cbind(chr = seqnames(cnvs[[x]]), 
                      as.data.frame(ranges(cnvs[[x]])), 
                      as.data.frame(mcols(cnvs[[x]]))) %>%
      unnest(cols = c(protein_name, starts_with("locus"), protein_product, starts_with("gene_"), length), keep_empty = FALSE) %>% 
      mutate(protein_name_short = str_squish(str_replace_all(str_remove_all(protein_name, '-like|isoform|LOW QUALITY PROTEIN:|X[0-9]|uncharacterized protein|"'), "F10 alpha chain", "F"))) %>%
      mutate(TUMOUR = parse_number(TUMOUR)) %>% mutate(id=row_number())

    message(paste("Running STRING on", x))
    
    map_to <- c("locus", "protein_name", "protein_name_short")
    cnv_data <- multi_map_df(as.data.frame(cnv_data),string_db$get_aliases(), map_to, "alias", "STRING_id", caseSensitive = FALSE)%>% left_join(colors,by="TUMOUR") %>% distinct()
    
    naDf <- subset(cnv_data, is.na(STRING_id))
    if (nrow(naDf) > 0) {
      cat(paste("Warning:  we couldn't map to STRING ", as.integer((nrow(naDf) / nrow(cnv_data)) *
        100), "% of your identifiers: ", sep = ""))
      cat(paste(head(naDf)$genes, sep = ", "))
    }
    

    save(cnv_data, file = paste0(saveloc, x, ".gz"), compress = TRUE)
  }
  if (params$reuse & file.exists(paste0(saveloc,  x, ".pdf"))) {
    message("Graph already exists.")
  } else {
    message(paste("Saving to", paste0(saveloc, x, ".pdf")))
    payload_id <- string_db$post_payload( cnv_data$STRING_id,
                                          colors=cnv_data$color )

    pdf(file = paste0(saveloc,  x, ".pdf"), width = 20, height = 18)
    tryCatch(string_db$plot_network(unique(cnv_data$STRING_id),payload_id=payload_id), error = function(cond) {
      # this likely happends if there's no valid STRING_ids
      # (ie, coulding map to STRING 100% of your identifiers)
      message(paste("Couldn't plot network of", count(map), "genes."))
      message(cond)
    })
    dev.off()
  }
  return(cnv_data)
}


mappings_file <- paste0(partialsloc, "mappings.gz")
if (params$reuse & file.exists(mappings_file)) {
  load(mappings_file)
} else {
  mappings <- lapply(unique(names), peformString, cnvs)
  names(mappings) <- unique(names)
  save(mappings, file = mappings_file)
}
```
### StringDB Plots
```{bash eval=FALSE, engine.path="/bin/bash", include=FALSE}
pwd
cd within/
mogrify -verbose -density 500 -resize 800 -format png ./*.pdf
```

## StringDB Enrichment 
We can then seperate them into areas with a copy number gain (CN > 2) or a copy number loss (CN < 1)
```{r}
mappings_gain <- lapply(names(mappings), function(name, mappings){
  gain <- mappings[[name]] %>% filter(TUMOUR > 2)
  print(paste0(count(gain), "/", count(mappings[[name]]), " (", round(100*count(gain)/count(mappings[[name]]),2), "%) of CNV regions in ", name, " were gains."))
  return(gain)
  },mappings)
mappings_loss <- lapply(names(mappings), function(name, mappings){
  loss <- mappings[[name]] %>% filter(TUMOUR < 2)
  print(paste0(count(loss), "/", count(mappings[[name]]), " (", round(100*count(loss)/count(mappings[[name]]),2), "%) of CNV regions in ", name, " were losses."))
  return(loss)
  },mappings)
```
And go through the enrichments accordingly:
```{r Retrieve Enrichments}
getEnrichments <- function(mapping, what="gain"){
  options(timeout=600) # if we're downloading stringdb, we need a longer time out 
  enrichments_file <- paste0(partialsloc, what, "-enrichments.gz")
  if(params$reuse && file.exists(enrichments_file)){
    load(enrichments_file) # load the enrichment file if relevant 
  } else {
  enrichments <- lapply(mapping, function(x){
    if(length(x$STRING_id %>% na.omit()) > 0) { # guard against empty mappings
       enrich <- string_db$get_enrichment( x$STRING_id ) %>% arrange(p_value)
       if(count(enrich) > 0){ # and empty enrichments 
         return(enrich)
       }
       message("No enrichments found.")
    }
    message("Returning NA.")
    return(NA)
  }) # otherwise get the enrichment information 
  # and remove any NAs 
  save(enrichments,file=enrichments_file) # and save it for the future 
  }
  return(enrichments)
}
enrichments <- getEnrichments(mappings, "all")
enrichments_gain <- getEnrichments(mappings_gain, "gain")
enrichments_loss <- getEnrichments(mappings_loss, "loss")
names(enrichments) <- names(enrichments_gain) <- names(enrichments_loss) <- names
```

### Enrichment Plots 
```{r UpSetR Settings}
colors_edge <- as.vector(c("#e41a1c", rep("#377eb8",2), "#4daf4a", "#984ea3","#c36100", "#c47800", "#c38e00", rep("#bfa407", 3), "#b9b925"))
colors_fill <- alpha(lighten(colors_edge,amount=2))
sets <- rev(names)
metadata <- as.data.frame(cbind(sets, colors=sets))

getEnrichTerms <- function(x){
  if(is.na(x)){
    return(x)
  }
  x <- x %>% filter(p_value < 0.05) %>% na.omit()
  return(x$term)}


upsetPlot <- function(terms,name="temp",what="terms"){
  try({
  plot <- upset(fromList(enrich_terms), order.by = "degree",mb.ratio = c(0.4, 0.6),mainbar.y.label =what, nintersects = NA, sets.x.label = paste(what,"per Sample"),sets=rev(names),keep.order = TRUE,sets.bar.color = rev(colors_edge), set.metadata = list(data = metadata, plots = list(list(type = "matrix_rows", 
    column = "colors", colors = setNames(colors_edge, names), 
    alpha = 1))))
  message(paste("Saving plot at", paste0(saveloc, name, ".pdf")))
  pdf(file=paste0(saveloc, name, ".pdf"),width=7,height=5)
  plot
  dev.off()
  return(plot)
  })
}

showRelevantTerms <- function(terms,enrichments){
  try({
  fromList(terms) %>% mutate(term=unique(unlist(terms))) %>%
  adorn_totals(where="col") %>% arrange(-Total)%>%
  left_join(enrichments[!is.na(enrichments)] %>% purrr::reduce(bind_rows) %>% dplyr::select(term, description), by =c("term" = "term")) %>% 
  distinct() 
  })
}
```

```{r Enrichment Term Overlap Gain, fig.height=7, fig.width=10}
enrich_terms <- lapply(enrichments_gain, getEnrichTerms)
names(enrich_terms) <- names # maintain the names 
upsetPlot(enrich_terms, "enrichment_gain_overlap", "Terms with CN gain")
head(showRelevantTerms(enrich_terms, enrichments_gain),n=30)
```     
```{r Enrichment Term Overlap Loss}
enrich_terms <- lapply(enrichments_loss, getEnrichTerms)
names(enrich_terms) <- names # maintain the names 
upsetPlot(enrich_terms, "enrichment_loss_overlap", "Terms with CN loss")
head(showRelevantTerms(enrich_terms, enrichments_loss),n=30)
```
```{r Enrichment Term Overlap Overall}
enrich_terms <- lapply(enrichments, getEnrichTerms)
names(enrich_terms) <- names 
upsetPlot(enrich_terms, "enrichment_all_overlap", "terms")
head(showRelevantTerms(enrich_terms, enrichments),n=30)
```

```{r Tidy Enrichment Data}
all_enrichments <- bind_rows(
  gain = bind_rows(enrichments_gain[!is.na(enrichments_gain)], .id = "sample"),
  loss = bind_rows(enrichments_loss[!is.na(enrichments_loss)], .id = "sample"), .id = "CN"
) %>%
  arrange(p_value) %>%
  filter(p_value < 0.05) %>%
  mutate(sample_short = str_trunc(sample, 4))

enrichment_stats <- all_enrichments %>%
  group_by(term) %>%
  summarize(count = n(), mean_p = mean(p_value), min_p = min(p_value), mean_gene = mean(number_of_genes), sum_gene = sum(number_of_genes))

all_enrichments <- all_enrichments %>%
  left_join(enrichment_stats) %>%
  distinct()
all_enrichments <- all_enrichments %>%
  left_join(sample_data, by = c("sample" = "sample_name")) %>%
  distinct()
```

```{r}
relevant <- fromList(enrich_terms) %>% mutate(enrich=unique(unlist(enrich_terms))) %>%
  mutate(all_yu = purrr::reduce(dplyr::select(., starts_with("yu")), `+`)) %>%
  dplyr::select(-starts_with("yu")) %>% mutate(all_yu = case_when(all_yu > 0 ~ 1, TRUE ~ 0)) %>% 
  adorn_totals(where="col") %>% arrange(-Total)%>%
  left_join(enrichments[!is.na(enrichments)] %>% purrr::reduce(bind_rows) %>% dplyr::select(term, description), by =c("enrich" = "term")) %>% 
  distinct() 

relevant %>% filter(Total > 1)
```


```{r}
all_enrichments <- all_enrichments %>% mutate(description = case_when(term == "PMID.32235722" ~ "(2020) T-Cells in Lymphoid Stress Surveillance", TRUE ~ str_trunc(description, 50) ))
theme_set(theme_minimal())

gplot <- ggplot(all_enrichments %>% filter(term %in% (relevant %>% filter(Total > 1))$enrich),
       aes(y=reorder(description,-count),x=number_of_genes))+
  geom_col(aes(fill=sample,color=sample)) +
  scale_fill_manual(values=setNames(colors_fill, names),labels=setNames(sample_data$sample_des, sample_data$sample_name)) +
  scale_color_manual(values=setNames(colors_edge, names),labels=setNames(sample_data$sample_des, sample_data$sample_name)) +
  labs(title="Common Enrichment Terms",y="Shared Terms between Turtles",x="Number of Genes") +
  xlim(c(0,40)) + theme(axis.text.x=element_text(angle=50,vjust=1,hjust=1),panel.grid.major.y = element_blank(),panel.grid.minor.x = element_blank())

try({gplot <- gplot +
  facet_wrap(vars(CN))})
gplot
ggsave(paste0(saveloc, "common_enrich_bar.pdf"),width = 7,height=5)


gplot <- ggplot(all_enrichments %>% filter(term %in% (relevant %>% filter(Total > 1))$enrich),
       aes(y=reorder(description,-count),x=p_value)) + geom_boxplot() + geom_point(aes(fill=sample,color=sample)) + scale_fill_manual(values=setNames(colors_fill, names),labels=setNames(sample_data$sample_des, sample_data$sample_name)) +
  scale_color_manual(values=setNames(colors_edge, names),labels=setNames(sample_data$sample_des, sample_data$sample_name)) +
  labs(title="Common Enrichment Terms",y="Shared Terms between Turtles",x="p-Value") +
  xlim(c(0,0.01)) + theme(axis.text.x=element_text(angle=50,vjust=1,hjust=1),panel.grid.major.y = element_blank(),panel.grid.minor.x = element_blank())

try({gplot <- gplot +
  facet_wrap(vars(CN))})
gplot
ggsave(paste0(saveloc, "common_enrich_box.pdf"),width = 7,height=5)
```

```{r}
tryCatch({
ggplot(all_enrichments %>% filter(term %in% head(enrichment_stats %>% arrange(mean_p),n=50)$term),aes(x=reorder(term,-mean_p),y=number_of_genes,fill=CN)) + geom_col() + labs(title="Common Enrichment Terms",x="Top 50 Enrichments by Mean P-Value",y="Number of Genes",fill="Copy\nNumber") + ylim(c(0,40)) + facet_grid( turtle ~ .,scales="free")+ theme(axis.text.x=element_text(angle=90,vjust=1,hjust=1))
ggsave(paste0(saveloc, "common_enrich_turtle.pdf"),width = 7,height=5)
},error=message)

tryCatch({
ggplot(all_enrichments %>% filter(term %in% head(enrichment_stats %>% filter(str_trunc(term,width=4,ellipsis="") != "PMID") %>%  arrange(mean_p),n=50)$term),aes(x=reorder(term,-mean_p),y=number_of_genes,fill=CN)) + geom_col() + labs(title="Common Enrichment Terms",x="Top 50 Enrichments by Mean P-Value",y="Number of Genes",fill="Copy\nNumber") +  facet_grid( turtle ~ .)+ theme(axis.text.x=element_text(angle=90,vjust=1,hjust=1))
ggsave(paste0(saveloc, "common_enrich_turtle_noPMID.pdf"),width = 7,height=5)
},error=message)
```


#### HSA-1280218
```{r}
tryCatch({
rel_row <- relevant[relevant$enrich == "HSA-1280218",] %>% na.omit()
enrich_common <- all_enrichments %>% filter(term == rel_row$enrich)
ggplot(enrich_common, aes(x=sample, y=number_of_genes,fill=sample)) + geom_col() + scale_fill_manual(values=setNames(colors_edge, names)[enrich_common$sample]) +  theme(legend.position = "none",panel.grid.major.x=element_blank()) + labs(title=paste(rel_row$enrich, "-", rel_row$description),x="",y="Number of Genes in Sample") + facet_wrap(vars(CN)) + scale_x_discrete(labels=setNames(sample_data$sample_des, sample_data$sample_name))+ theme(axis.text.x=element_text(angle=50,vjust=1,hjust=1))

enrich_common %>% dplyr::select(sample,preferredNames,p_value, CN)
ggsave(paste0(saveloc, "HSA-1280218.pdf"),width = 7,height=5)
},error=message)
```
#### SM00589 
```{r}
tryCatch({
rel_row <- relevant[relevant$enrich == "SM00589",] %>% na.omit()
enrich_common <- all_enrichments %>% filter(term == rel_row$enrich)
ggplot(enrich_common, aes(x=sample, y=number_of_genes,fill=sample)) + geom_col() + scale_fill_manual(values=setNames(colors_edge, names)[enrich_common$sample]) +  theme(legend.position = "none",panel.grid.major.x=element_blank()) + labs(title=paste(rel_row$enrich, "-", rel_row$description),x="",y="Number of Genes in Sample") + facet_wrap(vars(CN)) + scale_x_discrete(labels=setNames(sample_data$sample_des, sample_data$sample_name))+ theme(axis.text.x=element_text(angle=45,vjust=1,hjust=0.8))

enrich_common %>% dplyr::select(sample,preferredNames,p_value, CN)
ggsave(paste0(saveloc, "SM00589.pdf"),width = 7,height=5)


enrich_common %>% dplyr::select(sample,preferredNames,p_value, CN)
},error=message)
```
#### SM00449
```{r}
tryCatch({
rel_row <- relevant[relevant$enrich == "SM00449",]%>% na.omit()
enrich_common <- all_enrichments %>% filter(term == rel_row$enrich)
ggplot(enrich_common, aes(x=sample, y=number_of_genes,fill=sample)) + geom_col() + scale_fill_manual(values=setNames(colors_edge, names)[enrich_common$sample]) +  theme(legend.position = "none",panel.grid.major.x=element_blank()) + labs(title=paste(rel_row$enrich, "-", rel_row$description),x="",y="Number of Genes in Sample") +  scale_x_discrete(labels=setNames(sample_data$sample_des, sample_data$sample_name))+ theme(axis.text.x=element_text(angle=45,vjust=1,hjust=0.8))

enrich_common %>% dplyr::select(sample,preferredNames,p_value, CN)
ggsave(paste0(saveloc, "SM00449.pdf"),width = 7,height=5)


enrich_common %>% dplyr::select(sample,preferredNames,p_value, CN)
},error=message)
```

#### PF13765
```{r}
tryCatch({
rel_row <- relevant[relevant$enrich == "PF13765",]%>% na.omit()
enrich_common <- all_enrichments %>% filter(term == rel_row$enrich)
ggplot(enrich_common, aes(x=sample, y=number_of_genes,fill=sample)) + geom_col() + scale_fill_manual(values=setNames(colors_edge, names)[enrich_common$sample]) +  theme(legend.position = "none",panel.grid.major.x=element_blank()) + labs(title=paste(rel_row$enrich, "-", rel_row$description),x="",y="Number of Genes in Sample") + facet_wrap(vars(CN)) + scale_x_discrete(labels=setNames(sample_data$sample_des, sample_data$sample_name))+ theme(axis.text.x=element_text(angle=45,vjust=1,hjust=0.8))

enrich_common %>% dplyr::select(sample,preferredNames,p_value, CN)
ggsave(paste0(saveloc, "PF13765.pdf"),width = 7,height=5)

enrich_common %>% dplyr::select(sample,preferredNames,p_value, CN)
},error=message)
```

#### PMID.32235722
```{r}
tryCatch({
rel_row <- relevant[relevant$enrich == "PMID.32235722",]%>% na.omit()
enrich_common <- all_enrichments %>% filter(term == rel_row$enrich)
ggplot(enrich_common, aes(x=sample, y=number_of_genes,fill=sample)) + geom_col() + scale_fill_manual(values=setNames(colors_edge, names)[enrich_common$sample]) +  theme(legend.position = "none",panel.grid.major.x=element_blank()) + labs(title=paste(rel_row$enrich, "-", rel_row$description),x="",y="Number of Genes in Sample") +  scale_x_discrete(labels=setNames(sample_data$sample_des, sample_data$sample_name))+ theme(axis.text.x=element_text(angle=45,vjust=1,hjust=0.8))

enrich_common %>% dplyr::select(sample,preferredNames,p_value, CN)
ggsave(paste0(saveloc, "PMID.32235722.pdf"),width = 7,height=5)

enrich_common %>% dplyr::select(sample,preferredNames,p_value, CN)
},error=message)
```


## Euler Plots
Doing a euler or venn for all 12 samples is too much to accurately represent, but we can group them instead: 

### Group by Turtle 

```{r}
library(eulerr)
theme_set(theme_gray())
genes_by_turtle <- list(`27L`=enrich_terms$`27L1Fdna`, Flower=unique(c(enrich_terms$flSCEYfdna, enrich_terms$flSCINFdna)), Poppy=enrich_terms$poSCTFdna, Aladar=enrich_terms$`TABT-Cm`, Yucca=unique(c(enrich_terms$yuLIRSFdna, enrich_terms$yuRERFdna, enrich_terms$yuRIRSFdna, enrich_terms$yuRKTGFdna, enrich_terms$yuRKTMFdna, enrich_terms$yuRKTW1Fdna, enrich_terms$yuTSFdna)))
fit <- euler(genes_by_turtle,shape="ellipse")
plot <- plot(fit,fills=unique(colors_fill)[1:length(genes_by_turtle)],adjust_labels=TRUE,quantities=TRUE, legend=TRUE,edges=unique(colors_edge)[1:length(genes_by_turtle)])
plot
pdf(file = paste0(saveloc,  "-euler-turtle.pdf"), width = 7, height = 5)
plot
dev.off()

upset(fromList(genes_by_turtle), order.by = "degree",mb.ratio = c(0.4, 0.6),mainbar.y.label = "Signifigant terms", nintersects = NA, sets.x.label = "Signifigant terms per Turtle")
ggsave(paste0(saveloc, "-upsetr-turtle.pdf"),width = 7,height=5)
head(fromList(genes_by_turtle) %>% mutate(region=unique(unlist(genes_by_turtle))) %>%  adorn_totals(where="col") %>% arrange(-Total) %>% filter(Total > 1),40) %>% dplyr::select(region)
```
```{r}
genes_by_tissue <- list(Lung=unique(c(enrich_terms$`27L1Fdna`,enrich_terms$`TABT-Cm`)), Eye=unique(c(enrich_terms$flSCEYfdna,enrich_terms$yuRERFdna)), Inguinal=unique(c(enrich_terms$flSCINFdna,enrich_terms$yuLIRSFdna, enrich_terms$yuRIRSFdna)), Tail=unique(c(enrich_terms$poSCTFdna,enrich_terms$yuTSFdna)),Kidney=unique(c(enrich_terms$yuRKTGFdna, enrich_terms$yuRKTMFdna, enrich_terms$yuRKTW1Fdna)))
fit <- euler(genes_by_tissue,shape="ellipse")
plot <- plot(fit,adjust_labels=TRUE,quantities=TRUE, legend=TRUE)
plot
pdf(file = paste0(saveloc,  "-euler-tissue.pdf"), width = 7, height = 5)
plot
dev.off()

upset(fromList(genes_by_tissue), order.by = "degree",mb.ratio = c(0.4, 0.6),mainbar.y.label = "Signifigant terms", nintersects = NA, sets.x.label = "Signifigant terms per Tissue")
```

```{r}
genes_by_src <- list( External=unique(c(enrich_terms$flSCEYfdna,enrich_terms$yuRERFdna,enrich_terms$flSCINFdna,enrich_terms$yuLIRSFdna, enrich_terms$yuRIRSFdna,enrich_terms$poSCTFdna,enrich_terms$yuTSFdna)),Internal=unique(c(enrich_terms$yuRKTGFdna, enrich_terms$yuRKTMFdna, enrich_terms$yuRKTW1Fdna,enrich_terms$`27L1Fdna`,enrich_terms$`TABT-Cm`)))
fit <- euler(genes_by_src,shape="ellipse")
plot <- plot(fit,adjust_labels=TRUE,quantities=TRUE, legend=TRUE)
plot
pdf(file = paste0(saveloc,  "-euler-src.pdf"), width = 7, height = 5)
plot
dev.off()
```




